[IA-2088] [IA-2141] [IA-2089] [IA-2170] [IA-2140] Back Leo helm installs Galaxy

kyuksel opened the pull request:
This PR covers the following tickets:

* [IA-2088](https://broadworkbench.atlassian.net/browse/IA-2088) Back Leo installs Galaxy via Helm client lib
* [IA-2141](https://broadworkbench.atlassian.net/browse/IA-2141) Build string of override values
   * All known `--set` flags to the Galaxy install should now be present
* [IA-2089](https://broadworkbench.atlassian.net/browse/IA-2089) Update back Leo to uninstall charts
   * See https://github.com/broadinstitute/helm-scala-sdk/pull/5. 
   * The helm uninstall seems to work, but there is a known Galaxy issue with pods not terminating (Alex is looking into it)
* [IA-2170](https://broadworkbench.atlassian.net/browse/IA-2170) Add Role and Role Binding to helm install galaxy
   * This PR creates a KSA and passes it to `rbac` in the `galaxykubeman` chart
   * Note it does NOT set up Workload Identity yet (that will be a separate PR)
* [IA-2140](https://broadworkbench.atlassian.net/browse/IA-2140) Add code for monitoring waiting for app to come up
   * Added basic monitoring based on http polling. It works for now, but we can add k8s job polling in a future iteration.

Tested create/delete app flows in a fiab, and all seems to work.

kyuksel committed:
refactor

kyuksel committed:
shell

kyuksel committed:
galaxy app config updates

kyuksel committed:
various

kyuksel committed:
compiling

kyuksel committed:
existing units tests passing

kyuksel committed:
refactor

kyuksel committed:
fix some of the galaxy chart override mistakes

kyuksel committed:
fix ingress path

kyuksel committed:
galaxykubeman chart override value parsing seems to work now

rtitle commented:
Just scanned, but overall looks great! Did you get past the `invalid value` issue? Maybe I'll give this branch a shot too

kyuksel commented:
Status as of the late morning: App creation fails with the error below

```
org.broadinstitute.dsde.workbench.leonardo.monitor.PubsubKubernetesError: An error occurred with a kubernetes operation from source app during action createGalaxyApp.
 
Original message: release galaxy-rls failed, and has been uninstalled due to atomic being set:

Deployment.apps "galaxy-rls-workflow" is invalid:

[spec.template.spec.nodeSelector: Invalid value: "\"cloud.google.com/gke-nodepool\"": prefix part a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), 
spec.template.spec.nodeSelector: Invalid value: "\"cloud.google.com/gke-nodepool\"": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')]
```

Curiously, `helm install`ing on the command line using the values string produced by the PR code doesn't reproduce the issue.

Haven't been able to reproduce the error on the sbt console using the Helm client repo either; though, I'm not certain if the issue I run into there may be happening before the values are parsed.

kyuksel commented:
Current status: Chart override string parsing appears to work now. Install ongoing 🤞 

>Did you get past the invalid value issue?

Yes. A few fixes that made the difference are included in [this diff](https://github.com/DataBiosphere/leonardo/pull/1587/files/2d6fe589d1f92b66e17d80c1c37ca650330a108a..c06a06020112d48c2c535c7632d4c26e98b0dfcf).

kyuksel commented:
Chart values passed in look like this:

```
nfs.storageClass.name=nfs-galaxy-rls-3

cvmfs.repositories.cvmfs-gxy-data-galaxy-rls-3=data.galaxyproject.org

cvmfs.repositories.cvmfs-gxy-main-galaxy-rls-3=main.galaxyproject.org

cvmfs.cache.alienCache.storageClass=nfs-galaxy-rls-3

galaxy.persistence.storageClass=nfs-galaxy-rls-3

galaxy.cvmfs.data.pvc.storageClassName=cvmfs-gxy-data-galaxy-rls-3

galaxy.cvmfs.main.pvc.storageClassName=cvmfs-gxy-main-galaxy-rls-3

galaxy.nodeSelector.cloud\.google\.com/gke-nodepool=k780ba7c-e503-45e9-a66d-ad6229a1e557

nfs.nodeSelector.cloud\.google\.com/gke-nodepool=k780ba7c-e503-45e9-a66d-ad6229a1e557

galaxy.ingress.path=/proxy/google/v1/apps/h-proj-2020-08-27-kyuksel/galaxy-rls-3/galaxy

galaxy.ingress.annotations.nginx\.ingress\.kubernetes\.io/proxy-redirect-from=https://437046468.jupyter-dev.firecloud.org

galaxy.ingress.annotations.nginx\.ingress\.kubernetes\.io/proxy-redirect-to=https://leonardo-fiab.dsde-dev.broadinstitute.org:30443

galaxy.ingress.hosts[0]=437046468.jupyter-dev.firecloud.org

galaxy.configs.galaxy\.yml.galaxy.single_user=hermione.owner@test.firecloud.org,galaxy.configs.galaxy\.yml.galaxy.admin_users=hermione.owner@test.firecloud.org
```

kyuksel committed:
Fix ingress.path - use appName instead of release

kyuksel commented:
Status update: The `ingress.path` fix doesn't seem to have fixed the two Galaxy pods coming up on the console

![image](https://user-images.githubusercontent.com/5438223/91494332-a9305480-e886-11ea-9d55-02e789a5b669.png)

FWIW...also have been seeing the message `secrets "galaxy-rls-4-galaxy-secrets" not found` at the beginning of the installation although I can see that `secret` via `kubectl` on the command line  🤷‍♂️

kyuksel commented:
The final override values are below:

```
$ helm -n ky-glxy-app-0827-7-galaxy-ns get values galaxy-rls-4

USER-SUPPLIED VALUES:
cvmfs:
  cache:
    alienCache:
      storageClass: nfs-galaxy-rls-4
  repositories:
    cvmfs-gxy-data-galaxy-rls-4: data.galaxyproject.org
    cvmfs-gxy-main-galaxy-rls-4: main.galaxyproject.org
galaxy:
  configs:
    galaxy.yml:
      galaxy:
        admin_users: hermione.owner@test.firecloud.org
        single_user: hermione.owner@test.firecloud.org
  cvmfs:
    data:
      pvc:
        storageClassName: cvmfs-gxy-data-galaxy-rls-4
    main:
      pvc:
        storageClassName: cvmfs-gxy-main-galaxy-rls-4
  ingress:
    annotations:
      nginx.ingress.kubernetes.io/proxy-redirect-from: https://437046468.jupyter-dev.firecloud.org
      nginx.ingress.kubernetes.io/proxy-redirect-to: https://leonardo-fiab.dsde-dev.broadinstitute.org:30443
    hosts:
    - 437046468.jupyter-dev.firecloud.org
    path: /proxy/google/v1/apps/h-proj-2020-08-27-kyuksel/ky-glxy-app-0827-7/galaxy
  nodeSelector:
    cloud.google.com/gke-nodepool: k447d846-1c0a-4bd1-910d-65484f6809b8
  persistence:
    storageClass: nfs-galaxy-rls-4
nfs:
  nodeSelector:
    cloud.google.com/gke-nodepool: k447d846-1c0a-4bd1-910d-65484f6809b8
  storageClass:
    name: nfs-galaxy-rls-4
```

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/DateAccessedUpdater.scala:
```diff
@@ -32,7 +32,7 @@ class DateAccessedUpdater[F[_]: ContextShift: Timer](
 
   private[monitor] val check: F[Unit] =
-    logger.info(s"Going to update dateAccessed") >> queue.tryDequeueChunk1(config.maxUpdate).flatMap { chunks =>
+    logger.debug(s"Going to update dateAccessed") >> queue.tryDequeueChunk1(config.maxUpdate).flatMap { chunks =>
```
This was causing a lot of verbosity 🙈

Qi77Qi replied:
we can remove this line too...think I added this when I was debugging some issue

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -184,6 +186,8 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
         )
       )
 
+      _ <- logger.info(s"Successfully created cluster ${params.clusterId}!")
```
previously it was printing this message out late

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -456,6 +531,39 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
+
+    // Using the string interpolator raw""" since the chart keys include quotes to escape Helm
+    // value override special characters such as '.'
+    // https://helm.sh/docs/intro/using_helm/#the-format-and-limitations-of---set
```
`raw` interpolator may no longer be necessary since not using quotes in keys work the way we want. Even so, not sure if it buys anything to try to get rid of it 🤷‍♂️

kyuksel commented:
Also just noticed the following on my test cluster

![image](https://user-images.githubusercontent.com/5438223/91498246-7178db00-e88d-11ea-9900-1b85ee4d0548.png)


which may explain `galaxy-rls-4-cvmfs-cache-preload-0` being `Unschedulable` (see previous screenshot above), which might in turn have prevented the web pod from coming up.

rtitle committed:
Minor fixes, going to test in a fiab

rtitle committed:
Add basic galaxy monitoring

rtitle committed:
Release name must be unique per app within a cluster

rtitle commented:
FYI this all worked and I demoed in in the IA sprint review! I still want to clean some stuff up next week, and then I'll move this to Ready for review.

kyuksel commented:
> FYI this all worked and I demoed in in the IA sprint review! I still want to clean some stuff up next week, and then I'll move this to Ready for review.

Yay! Thank you 😊

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -379,12 +381,12 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
 
       _ <- logger.info(
-        s"Installing ingress helm chart '${config.ingressConfig.chart}' in cluster '${dbCluster.id}' | trace id: ${ctx.traceId}"
+        s"Installing ingress helm chart ${config.ingressConfig.chart} in cluster ${dbCluster.id} | trace id: ${ctx.traceId}"
```
Nit: Values read better to me in logs (i.e. `chart.value` instead of `chart`, and similarly for `id`)

rtitle replied:
+1 me too - will add `.value`. 

Also maybe it's just me but I really don't like quoted strings in logs like `s"Installing ingress helm chart '${config.ingressConfig.chart}'"`. Not sure why.

kyuksel replied:
Single quotes in log messages allow my brain to pause, like with `,` and easily identify that it's a value and not a regular word in the sentence. That being said, it's a personal preference and I don't expect it to work the same way for others 😊

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -472,14 +474,24 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
 
+      // Poll galaxy until it starts up
+      _ <- (
+        Stream.sleep_(config.monitorConfig.createApp.interval) ++
+          Stream.eval(
+            galaxyDAO.isProxyAvailable(dbCluster.googleProject, appName)
+          )
+      ).repeatN(config.monitorConfig.createApp.maxAttempts)
+        .takeThrough(_ == false)
+        .compile
+        .lastOrError
```
This pattern looks nice and seems to be utilized in a few places so it may be nice to factor it out into a util method that accepts parameters for `Stream.eval()`, `repeatN()` and `takeThrough()`.

rtitle replied:
Maybe… although I feel like the fs2 combinators are already pretty general and easy to read, not sure a util method would make things any simpler.

Qi77Qi replied:
you can use `streamFUntilDone` directly I think... you just need to provide a `DoneCheckable[Boolean]` instance

jdcanas replied:
you could use streamFUntilDone https://github.com/broadinstitute/workbench-libs/blob/develop/google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/package.scala#L121. Though it may be worth adding a param for onTimeout to it...

kyuksel reviewed (commented):
Looks great 👍 A couple of minor comments.

Also, what was the value string fix that made it work?

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -405,6 +434,65 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
+        )
+        .run(helmAuthContext)
+
+      // Poll galaxy until it starts up
```
Do we need a TODO to check the applicable Galaxy pod(s) as well when the dependencies are in place?

rtitle replied:
Yeah - I added a TODO in `HttpGalaxyDAO` but maybe it belongs here.

rtitle commented:
> Also, what was the value string fix that made it work?

I thought it was something around `galaxy.ingress.path` or `galaxy.ingress.host` but looking at commits I'm not sure. Another I found was that releaseName apparently needs to be unique within a cluster (not sure why) so I made it `releaseNameSuffix` in config.

rtitle committed:
TEMP update go lib

rtitle committed:
- add config maps to Galaxy install
- improve logging in GKEInterpreter
- add trailing slash to HttpGalaxyDAO to avoid 302 when status checking

rtitle committed:
Added helm delete galaxy, and improved helm monitoring a bit

rtitle committed:
Removed by mistake

rtitle committed:
Fix unit test failures

rtitle committed:
Create KSA and pass it to the helm command since it appears to be required now

Annotating for workload identity not implemented yet

rtitle committed:
- add customEnvironmentVariables to Swagger and GetAppResponse
- added keepHistory as a configurable param to helm uninstall

rtitle commented on Dockerfile:
```diff
@@ -9,7 +9,7 @@ RUN mkdir /helm-go-lib-build && \
     git clone https://github.com/broadinstitute/helm-scala-sdk.git && \
     cd helm-scala-sdk && \
-    git checkout master && \
+    git checkout rt-uninstall && \
```
TODO change back to master once https://github.com/broadinstitute/helm-scala-sdk/pull/5 is merged

rtitle commented on project/Dependencies.scala:
```diff
@@ -26,7 +26,7 @@ object Dependencies {
   val workbenchErrorReportingV = "0.1-92fcd96"
 
-  val helmScalaSdkV = "0.0.1-RC1"
+  val helmScalaSdkV = "0.0.1-rt-uninstall-SNAPSHOT"
```
TODO: change back to 0.0.1-RC1 once https://github.com/broadinstitute/helm-scala-sdk/pull/5 is merged

rtitle commented on http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreterSpec.scala:
```diff
@@ -12,6 +12,7 @@ import org.broadinstitute.dsde.workbench.google2.mock.{
 import org.broadinstitute.dsde.workbench.leonardo.CommonTestData._
```
TODO: add more specs to `GKEInterpreterSpec`

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/GalaxyAppConfig.scala:
```diff
@@ -1,6 +1,11 @@
+                           chart: ChartName,
+                           namespaceNameSuffix: NamespaceName,
+                           services: List[ServiceConfig],
+                           serviceAccountSuffix: KubernetesServiceAccount,
```
is this just suffix or a complete ksa?

rtitle replied:
Just a suffix - I'm generating the ksa as `${appName}-${serviceAccountSuffix}`

rtitle replied:
Maybe the suffixes should be `String`s? 🤔

Qi77Qi replied:
I think `String` sounds good to me...there's no other `String` param in this data type anyway, so there's no chance we'll mix it up with another param

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -264,23 +279,69 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
+      _ <- kubeService.createServiceAccount(gkeClusterId, ksa, KubernetesNamespace(namespaceName))
+
+      // update KSA in DB
+      _ = appQuery.updateKubernetesServiceAccount(app.id, ksaName).transaction
```
should this be `_ <-`? if true, probably should add unit test for this

rtitle replied:
Ahh good catch, should be `<-`. I'll try to add more specs to `GKEInterpreterSpec`

rtitle replied:
Added assertions to `LeoPubsubMessageSubscriberSpec` that the KSA is persisted to the DB

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -324,23 +387,44 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
+      dbCluster = dbApp.cluster
+      gkeClusterId = dbCluster.getGkeClusterId
+
+      _ <- logger.info(
```
should this be `debug`?

rtitle replied:
I dunno, I added a lot of `info` statements to make it clear what's going on when looking at logs. I feel like it's helpful.

kyuksel replied:
I find them helpful as well.

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -400,11 +471,157 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
+
+      helmAuthContext <- getHelmAuthContext(googleCluster, dbCluster.id, namespaceName)
+
+      releaseName = ReleaseName(s"${appName.value}-${config.galaxyAppConfig.releaseNameSuffix.value}")
```
would it make sense to suffix a timestamp in the release name?

rtitle replied:
I think it would - especially as we currently have the `keepHistory` option when deleting charts. Will update.

rtitle replied:
Hm.. we need `releaseName` at both install and delete time, so if we add a timestamp then we should probably persist it to the database. This may be a good idea anyway, but I just added a TODO for now.

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -400,11 +471,157 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
+        .uninstall(org.broadinstitute.dsp.Release(releaseName.value), config.galaxyAppConfig.uninstallKeepHistory)
+        .run(helmAuthContext)
+
+      isDone <- (Stream.sleep_(config.monitorConfig.deleteApp.interval) ++
```
same. you should be able to use `streamFUntilDone` and provide a `DoneCheckable[List[Pod]]` instance (hand wavy signature)

rtitle replied:
Forgot about `streamFUntilDone`! Thanks

rtitle committed:
PR feedback so far; still need to add unit tests

rtitle committed:
factor out buildReleaseName method

rtitle committed:
Fix DB call

rtitle commented:
Thanks @Qi77Qi -- applied feedback so far I believe, still need to add some `GKEInterpreterSpec` tests

rtitle committed:
helmScalaSdkV = 0.0.1-RC2

kyuksel commented on http/src/main/resources/reference.conf:
```diff
@@ -241,6 +241,8 @@ gke {
     chart = "galaxy/galaxykubeman"
     namespaceNameSuffix = "galaxy-ns"
+    serviceAccountSuffix = "galaxy-ksa"
+    uninstallKeepHistory = true
```
might want to consider adding a comment clarifying what history it is we're keeping

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -59,19 +62,20 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
         clusterOpt,
         KubernetesClusterNotFoundException(
-          s"Failed kubernetes cluster creation. Cluster with id ${params.clusterId} not found in database"
+          s"Failed kubernetes cluster creation. Cluster with id ${params.clusterId.id} not found in database | trace id: ${ctx.traceId}"
```
👍

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -59,19 +62,20 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
 
+      _ <- logger.info(
+        s"Beginning cluster creation for cluster ${dbCluster.getGkeClusterId.toString} | trace id: ${ctx.traceId}"
+      )
```
Should we say "...for cluster with ID ${dbCluster.getGkeClusterId.toString}..."? Similarly below.

rtitle replied:
`getGkeClusterId.toString` returns a string like `s"projects/${project.value}/locations/${location.value}/clusters/${clusterName.value}"`. I like it because it's pretty imformative in logging, but because of this I don't feel that including "with ID" is necessary. (Perhaps `getGkeClusterId` itself should be named differently).

kyuksel replied:
Oh right. No need for `with ID` then indeed. 

> Perhaps getGkeClusterId itself should be named differently)

Yea that and `KubernetesClusterId` seem misleading since that's not usually what we mean by `id`s in other places.

rtitle replied:
I feel like there are several places where we might want to refactor the Leo/wb-libs k8s models but I'm holding off.. maybe can do as tech debt once the alpha release is stable.

kyuksel replied:
Sounds good. Could add a TODO for future reference.

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GKEInterpreter.scala:
```diff
@@ -414,53 +458,140 @@ class GKEInterpreter[F[_]: Parallel: ContextShift: Timer](
+
+        // TODO (RT): currently there seems to be a bug causing Galaxy pods to not terminate when `helm delete` is called.
+        // So we're currently _not_ throwing an error if deletion times out, but we should uncomment the below
+        // line once the Galaxy bug is fixed.
```
Do we know if those `Terminating` pods are removed once the namespace is deleted?

rtitle replied:
They are not - the namespace itself gets stuck in Terminating and the pods also stay in Terminating.

It sounded like Alex had an idea about what the issue is.

kyuksel reviewed (commented):
The past two days' changes look good. Minor questions and comments are inline.

rtitle committed:
More PR feedback and add a few tests
